eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -213,7 +213,10 @@ object OutgoingPaymentPacket {
    *         - the trampoline onion to include in final payload of a normal onion
    */
   def buildTrampolineToLegacyPacket(invoice: PaymentRequest, hops: Seq[NodeHop], finalPayload: PaymentOnion.FinalPayload): (MilliSatoshi, CltvExpiry, Sphinx.PacketAndSecrets) = {
-    val (firstAmount, firstExpiry, payloads) = hops.drop(1).reverse.foldLeft((finalPayload.amount, finalPayload.expiry, Seq[PaymentOnion.PerHopPayload](finalPayload))) {
+    // NB: the final payload will never reach the recipient, since the next-to-last trampoline hop will convert that to a legacy payment
```

Comment:
The node just before the last node.
Is there a different comment that would make more sense to you?
I looked up the definition of "penultimate", but it says that it means "next to last" so I guess this should be more understandable?

Reply:
I think my confusion mainly comes from the term "hop". A hop is between two nodes but you actually mean a specific node. And I don't think it's the penultimate trampoline node that does the conversion, it's the last trampoline node, isn't it?

Reply:
Right, I will replace that with node, it makes more sense.

It is the penultimate trampoline node that does the conversion, not the last one, because for now the last node is the recipient: we build a trampoline onion for the complete route to the recipient (even though we know it will be converted to a legacy payment instead of being relayed). This was mostly done to minimize code changes, because I hoped we would be able to get rid of the "trampoline to legacy" case quickly (I thought wallet would take less time to at least implement the trampoline onion decryption, but that hasn't seen a lot of traction yet...).

Reply:
Clarified in https://github.com/ACINQ/eclair/pull/2063/commits/fa82531fa61b6728c84b28212fdcfeaa0b02d4fd

We should change that behavior some day, but I've been delaying that to when we implement the "new" version of trampoline (where trampoline onion is variable-size, as the latest state of the spec PR recommends).

Reply:
I wouldn't consider the last node to be a trampoline node. You can call it "the last trampoline node before the recipient" to remove all ambiguity.

Reply:
What's important here is that the recipient receives a trampoline payload in the onion, if that wasn't the case the comment would be inaccurate (or worse, the code would be incorrect). I used "trampoline node" to refer to "node that will receive a payload in the trampoline onion" for simplicity, I'll find a better way of expressing that.